bullet-featherstone/src/SDFFeatures.cc

```diff
@@ -0,0 +1,749 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include "SDFFeatures.hh"
+#include <gz/math/eigen3/Conversions.hh>
+#include <gz/math/Helpers.hh>
+
+#include <sdf/Link.hh>
+#include <sdf/Joint.hh>
+#include <sdf/Geometry.hh>
+#include <sdf/Box.hh>
+#include <sdf/Capsule.hh>
+#include <sdf/Ellipsoid.hh>
+#include <sdf/Sphere.hh>
+#include <sdf/Cylinder.hh>
+#include <sdf/Plane.hh>
+#include <sdf/JointAxis.hh>
+#include <sdf/Surface.hh>
+
+#include <BulletDynamics/Featherstone/btMultiBodyLinkCollider.h>
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace gz {
+namespace physics {
+namespace bullet_featherstone {
+
+/////////////////////////////////////////////////
+/// \brief Resolve the pose of an SDF DOM object with respect to its relative_to
+/// frame. If that fails, return the raw pose
+static std::optional<Eigen::Isometry3d> ResolveSdfPose(
+  const ::sdf::SemanticPose &_semPose)
+{
+  math::Pose3d pose;
+  ::sdf::Errors errors = _semPose.Resolve(pose);
+  if (!errors.empty())
+  {
+    if (!_semPose.RelativeTo().empty())
+    {
+      gzerr << "There was an error in SemanticPose::Resolve:\n";
+      for (const auto &err : errors)
+      {
+        gzerr << err.Message() << std::endl;
+      }
+    }
+
+    return std::nullopt;
+  }
+
+  return math::eigen3::convert(pose);
+}
+
+/////////////////////////////////////////////////
+Identity SDFFeatures::ConstructSdfWorld(
+    const Identity &_engine,
+    const ::sdf::World &_sdfWorld)
+{
+  const Identity worldID = this->ConstructEmptyWorld(_engine, _sdfWorld.Name());
+
+  const WorldInfoPtr &worldInfo = this->worlds.at(worldID);
+
+  auto gravity = _sdfWorld.Gravity();
+  worldInfo->world->setGravity(btVector3(gravity[0], gravity[1], gravity[2]));
+
+  for (std::size_t i = 0; i < _sdfWorld.ModelCount(); ++i)
+  {
+    const ::sdf::Model *model = _sdfWorld.ModelByIndex(i);
+
+    if (!model)
+      continue;
+
+    this->ConstructSdfModel(worldID, *model);
```

Comment:
this call may return an invalid ID, we should handle this, for example when a joint type is not supported we will get a crash